README.md

```diff
@@ -291,173 +321,205 @@ be used in global variable initializers.
-with the `string.advance`, `string.rewind`, `string.cur`,
-`string.measure`, `string.encode`, and `string.eq` instructions must
-trap.
+Evaluation of an instruction which would result in a WTF-16 position
+value greater than 2<sup>31</sup> must trap.  In practice, no web
+browser embedding allows for strings longer than 2<sup>31</sup>–1 WTF-16
+code units, so no string from JavaScript is too large for the
+instructions in this proposal, when processed as WTF-16.
 
-Given that there must be a distinct cursor value for each codepoint in a
-string, and one for the end, this constrains the strings that are
-processed by to a maximum of 2<sup>31</sup>–1 code units.  Not all
-strings have one code unit per codepoint, so the codepoint size limit in
-practice may be lower for any given string.  This specification can
-therefore represent codepoint counts with an `i32` without risk of
-overflow.
+Evaluation of an instruction which would result in a WTF-8 position
+value greater than 2<sup>31</sup> must trap.  Note that depending on the
+string contents, a position for a WTF-8 encoding may be numerically
+lesser or greater than the corresponding WTF-16 position.
 
 Future extensions of the `stringref` proposal along the lines of the
 [memory64
 proposal](https://github.com/WebAssembly/memory64/blob/main/proposals/memory64/Overview.md)
-may allow for 64-bit variants of the cursor-using instructions, which
+may allow for 64-bit variants of the position-using instructions, which
 could relax these restrictions.
 
-In practice, no web browser embedding allows for strings longer than
-2<sup>31</sup>–1 UTF-16 code units, so no string from JavaScript is too
-large for the instructions in this proposal.
-
 ### Accessing string contents
 
 All parameters and return values measuring a number of codepoints or a
-number of bytes represent these sizes as unsigned values.
+number of code units represent these sizes as unsigned values.
 
 ```
-(string.cur str:stringref cursor:i32)
+(string.get_wtf8 str:stringref pos:i32)
   -> codepoint:i32
 ```
-Return the codepoint at *`cursor`* in *`str`*, as an i32.  The result is
-a unicode scalar value, unless the host's strings support isolated
-surrogates, in which case the result may be a surrogate.  If *`cursor`*
-is at the end of *`str`*, the result is -1.
+Return the codepoint at the WTF-8 *`pos`* in *`str`*, as an i32.  The
+result will be a valid codepoint, and may be an isolated surrogate.  If
+*`pos`* is at the end of *`str`*, the result is -1.
 
 ```
-(string.measure $encoding str:stringref cursor:i32 codepoints:i32)
+(string.get_wtf16 str:stringref pos:i32)
+  -> codeunit:i32
+```
+Return the WTF-16 code unit at the WTF-16 *`pos`* in *`str`*, as an i32.
+The result may be part of a surrogate pair, or may be an isolated
+surrogate; the caller is responsible for decoding to codepoints as
+needed.  If *`pos`* is greater than or equal to the result of
+`string.end_wtf16` for *`str`*, trap.
```

Comment:
Why the apparent mismatch between returning -1 for wtf8 and trapping here?

Reply:
Yeah this is a weirdness: both `get_wtf8` and `get_wtf16` trap for bad positions.  However it's easy to characterize the set of good position values for `get_wtf16` -- it's the integers from 0 to the `string.end_wtf16`.  For `string.get_wtf8` you have to use `string.advance_wtf8` to iterate over them.

I was thinking that for WTF-16, what you would want to do is get the total UTF-16 length and in that way it would be easy to avoid the "end position", and therefore we know that `get_wtf16` can return a values in `[0,0xffff]`.  But for UTF-8 you wouldn't necessarily want to use `string.advance_wtf8` to get the end position (also getting number of codepoints then throwing it away); rather you would just want to iterate via `advance_wtf8` and get an in-band sentinel from `get_wtf8` when done.

WDYT?

Reply:
Updated to make these match.